source/mainnet/smart-contracts/guides/factory-pattern.rst

```diff
@@ -0,0 +1,392 @@
+
+The ``factory`` contract
+------------------------
+
+For the purposes of this example, the factory contract assigns each new product a unique (sequential)
+index. It also retains a map of all of the products that it has previously produced.
+The state of the factory contract is represented by the following Rust struct:
+
+.. code-block:: Rust
+
+    pub struct FactoryState<S: HasStateApi = StateApi> {
+        /// The index that will be assigned to the next product contract.
+        next_product: u64,
+        /// Index of the product smart contract instances.
+        products:     StateMap<u64, ContractAddress, S>,
+    }
+
+The ``produce`` endpoint
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``produce`` method of the factory expects one parameter that is the address of an uninitialized
+instance of the ``product`` contract. First of all, the parameter is read from the context:
+
+.. code-block:: Rust
+
+        let product_address = ctx.parameter_cursor().get()?;
+
+The factory needs to be sure that the address does actually refer to an instance of the ``product``
+contract, in order to ensure correct behavior. This can be achieved by checking the module reference
+and contract name against expected values. Together, the module reference and contract name uniquely
+identify the code of the smart contract instance.
+
+Getting the module reference and contract name is done using the host functions
+``contract_module_reference`` and ``contract_name``, respectively. Note: both of these functions
+are introduced in protocol version 7, and will not work while the chain is running an earlier
+protocol version.
+
+In this example, the factory and product contracts are defined in the same module.
+Thus, to check that the module reference of the product is correct, it is sufficient to check that
+it is the same as the module reference of the factory, as follows:
+
+.. code-block:: Rust
+
+        // We can depend upon getting the module reference for our own contract.
+        let self_module_ref = host.contract_module_reference(ctx.self_address()).unwrap();
+        // Check the product module is the same as our own module.
+        let product_module_ref = host
+            .contract_module_reference(product_address)
+            .or(Err(FactoryError::NonExistentProduct))?;
+        ensure_eq!(self_module_ref, product_module_ref, FactoryError::InvalidProduct);
+
+If the product were instead defined in a different module from the factory, then the expected
+module reference for the product could be hard-coded into the factory. Alternatively, if the
+module rereference for the product is to be determined later, it could be passed as a parameter
+when creating the ``factory`` instance.
+
+If the module defining the product is known to only contain one smart contract, then checking the
+module reference is sufficient for identifying the code of the product smart contract. In this case,
```

Comment:
```suggestion
module reference is sufficient for identifying the code of the product smart contract. In the example smart contract,
```

People could think that `in this case` refers to the example you describe in the previous sentence.